### PR TITLE
Adding preMigration option to  validate API definitions

### PR DESCRIPTION
**Pull request by HiranyaKavishani**
## Purpose
$Subject

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java [hunk 1]
```diff
@@ -116,32 +143,43 @@ private void populateApplicationOrganizations() throws APIMigrationException {
...
+        String formattedTenantConf = null;
+
+        try {
+            if (tenantConf != null) {
+                tenantConf.putIfAbsent(Constants.IS_UNLIMITED_TIER_PAID, false);
+                formattedTenantConf = mapper.writerWithDefaultPrettyPrinter().writeValueAsString(tenantConf);
             }
+        } catch (JsonProcessingException jse) {
+            log.info("Error while JSON Processing tenant conf :" + jse);
+            log.info("Hence, skipping tenant conf to db migration for tenant Id :" + tenantId);
```
Comment (rmsamitha) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:165]: Thes should be error logs (no info).
No need to add two logging lines for this as the logged content can be mention in a single log entry.

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java [hunk 2]
```diff
@@ -116,32 +143,43 @@ private void populateApplicationOrganizations() throws APIMigrationException {
...
-                log.info(
-                        "Error while adding to tenant conf to database for tenant: " + tenantId + "with Error :" + ape);
-                continue;
+            } catch (APIManagementException e) {
+                log.info("Error while adding to tenant conf to database for tenant: " + tenantId + "with Error :" + e);
             }
-        }
+        } else {
+             if (log.isDebugEnabled()) {
+                 log.debug("tenant conf value is empty");
```
Comment (rmsamitha) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:177]: This also is an error AFAIU. 
User should be informed on this without enabling debug logs.

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java [hunk 3]
```diff
@@ -301,6 +339,289 @@ public void registryDataPopulation() throws APIMigrationException {
         log.info("Rxt resource migration done for all the tenants");
     }
 
+    /**
+     *  This method can be used as prevalidation step
```
Comment (rmsamitha) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:343]: Empty line missing after the method description

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java [hunk 4]
```diff
@@ -301,6 +339,289 @@ public void registryDataPopulation() throws APIMigrationException {
...
+     * @param validateStep validateStep
+     * @throws APIMigrationException
```
Comment (rmsamitha) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:345]: No use of adding doc comment lines without adding a description. If there is no specific description for this throws clause, better remove the doc comment line

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java [hunk 5]
```diff
@@ -301,6 +339,289 @@ public void registryDataPopulation() throws APIMigrationException {
...
+        ObjectMapper mapper = new ObjectMapper();
+        if (preValidationServiceList.contains(validateStep)) {
+            for (Tenant tenant : getTenantsArray()) {
+                JSONObject tenantConf = getTenantConfigFromRegistry(tenant.getId());
+                tenantConf.putIfAbsent(Constants.IS_UNLIMITED_TIER_PAID, false);
+                String formattedTenantConf = null;
+                try {
+                    formattedTenantConf = mapper.writerWithDefaultPrettyPrinter().writeValueAsString(tenantConf);
+                } catch (JsonProcessingException e) {
+                    log.error("Error while JSON Processing tenant conf :" + e);
```
Comment (rmsamitha) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:360]: Don't we need to stop the code flow after this error occurred. (by throwing exception)

**Review by rmsamitha — CHANGES_REQUESTED** [review]

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrationClientBase.java [hunk 6]
```diff
@@ -231,6 +231,16 @@ public void doMigration(TreeMap<String, MigrationClient> migrationServiceList, S
         }
     }
 
+    public void doValidation(TreeMap<String, MigrationClient> migrationServiceList, String runPreMigrationStep)
+            throws APIMigrationException {
+        log.info("Pre Migration step to validate data is executing ..........");
```
Comment (rmsamitha) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrationClientBase.java:236]: Should be "Executing the Pre Migration step to validate data .........."

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java [hunk 7]
```diff
@@ -301,6 +339,289 @@ public void registryDataPopulation() throws APIMigrationException {
...
+     *  This method can be used as prevalidation step
+     * @param validateStep validateStep
+     * @throws APIMigrationException
+     */
+    @Override
+    public void preMigrationValidation(String validateStep) throws APIMigrationException {
+
+        boolean isTenantFlowStarted = false;
+        ObjectMapper mapper = new ObjectMapper();
+        if (preValidationServiceList.contains(validateStep)) {
```
Comment (rmsamitha) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:352]: Exit the method if this condition is false.

Using long body in if bliock makes the code less readble.

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java [hunk 8]
```diff
@@ -301,6 +339,289 @@ public void registryDataPopulation() throws APIMigrationException {
...
+                        PrivilegedCarbonContext.endTenantFlow();
+                    }
+                }
+                if (log.isDebugEnabled()) {
+                    log.debug("End pre migration validation of api definition for tenant " + tenant.getId() + '(' + tenant
+                            .getDomain() + ')');
+                }
+            }
+        } else {
+            log.info("Pre migration step" + validateStep + " is not defined.");
```
Comment (rmsamitha) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:391]: Note the above comment too.
And isn't this log should be an error log or at least a WARN? (not sure how we config this configuration)

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrationClientBase.java (hunk 6 above)
Comment (HiranyaKavishani) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrationClientBase.java:236]: +1

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java [hunk 9]
```diff
@@ -301,6 +339,289 @@ public void registryDataPopulation() throws APIMigrationException {
...
+                        if (artifactPath.contains("/apimgt/applicationdata/apis/")) {
+                            continue;
+                        }
+
+                        API api = APIUtil.getAPI(artifact, registry);
+
+                        // validate API Definitions
+                        validateAPIDefinitions(api, registry);
+                    } catch (Exception e) {
+                        throw new APIMigrationException("Error occurred while retrieving API from the registry: ", e);
```
Comment (rmsamitha) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:423]: Mention the api artifact path/name here

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java (hunk 1 above)
Comment (HiranyaKavishani) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:165]: Previous code is moved to separate method.+1

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java [hunk 10]
```diff
@@ -301,6 +339,289 @@ public void registryDataPopulation() throws APIMigrationException {
...
+                log.info("Starting validate the api definitions of tenant " + tenant.getDomain() + "..........");
+
+                for (GenericArtifact artifact : artifacts) {
+                    try {
+                        String artifactPath = ((GenericArtifactImpl) artifact).getArtifactPath();
+                        if (log.isDebugEnabled()) {
+                            log.debug("artifact path:  " + artifactPath );
+                        }
+
+                        if (artifactPath.contains("/apimgt/applicationdata/apis/")) {
```
Comment (rmsamitha) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:414]: What is the requirement of this check?

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java (hunk 8 above)
Comment (HiranyaKavishani) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:391]: It is just like we do in migrationFromVersion

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java [hunk 11]
```diff
@@ -301,6 +339,289 @@ public void registryDataPopulation() throws APIMigrationException {
...
+        if (log.isDebugEnabled()) {
+            log.debug("Validating open api definition of  " + api.getId().getApiName() + " version: " + api.getId()
+                    .getVersion() + "type: " + api.getType());
+        }
+
+        try {
+            String swaggerDefinition = OASParserUtil.getAPIDefinition(api.getId(), registry);
+            validationResponse = OASParserUtil.validateAPIDefinition(swaggerDefinition, Boolean.TRUE);
+        } catch (APIManagementException e) {
+            log.error(" Error while validating open api definition. " + e);
```
Comment (rmsamitha) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:476]: Mention the api name , version, tenent, etc

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java [hunk 12]
```diff
@@ -301,6 +339,289 @@ public void registryDataPopulation() throws APIMigrationException {
...
+        }
+
+        if (validationResponse != null && !validationResponse.isValid()) {
+            for (ErrorHandler error : validationResponse.getErrorItems()) {
+                log.error("Open API Definition is invalid. ErrorMessage: " + error.getErrorMessage()
```
Comment (rmsamitha) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:481]: Mention the api name , version, tenent, etc

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java [hunk 13]
```diff
@@ -301,6 +339,289 @@ public void registryDataPopulation() throws APIMigrationException {
...
+                String graphqlSchema = definition.getGraphqlSchemaDefinition(api.getId(), registry);
+                graphQLValidationResponseDTO = PublisherCommonUtils
+                        .validateGraphQLSchema("schema.graphql", graphqlSchema);
+            } catch (APIManagementException e) {
+                log.error(" Error while validating graphql api definition. " + e);
+            }
+        }
+
+        if (graphQLValidationResponseDTO != null && !graphQLValidationResponseDTO.isIsValid()) {
+            log.error(" Invalid GraphQL definition found. " + "ErrorMessage: " + graphQLValidationResponseDTO
```
Comment (rmsamitha) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:513]: Mention the api name , version, tenent, etc if the ErrorMessage doesn't include them

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java [hunk 14]
```diff
@@ -301,6 +339,289 @@ public void registryDataPopulation() throws APIMigrationException {
...
+            }
+        } catch (RegistryException e) {
+            log.error(" Error while getting wsdl file. " + e);
+        } catch (APIManagementException e) {
+            log.error(" Error while validating wsdl file. " + e);
+        }
+
+
+        if (wsdlValidationResponse != null && !wsdlValidationResponse.isValid()) {
+            log.error(" Invalid WSDL definition found. " + wsdlValidationResponse.getError());
```
Comment (rmsamitha) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:557]: Mention the api name , version, tenent, etc if the ErrorMessage doesn't include them

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java [hunk 15]
```diff
@@ -301,6 +339,289 @@ public void registryDataPopulation() throws APIMigrationException {
...
+                        apiSourcePath + RegistryConstants.PATH_SEPARATOR + api.getId().getProviderName() + "--" + api
+                                .getId().getApiName() + api.getId().getVersion() + ".wsdl";
+                wsdl = (byte[]) registry.get(wsdlResourcePath).getContent();
+                wsdlValidationResponse = APIMWSDLReader.validateWSDLFile(wsdl);
+            }
+        } catch (RegistryException e) {
+            log.error(" Error while getting wsdl file. " + e);
+        } catch (APIManagementException e) {
+            log.error(" Error while validating wsdl file. " + e);
+        }
```
Comment (rmsamitha) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:553]: Mention the api name , version, tenent, etc

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java [hunk 16]
```diff
@@ -301,6 +339,289 @@ public void registryDataPopulation() throws APIMigrationException {
...
+
+        if (log.isDebugEnabled()) {
+            log.debug("Validating streaming api definition of " + api.getId().getApiName() + " version: "
+                    + api.getId().getVersion() + "type: " + api.getType());
+        }
+
+        try {
+            apiPath = GovernanceUtils.getArtifactPath(registry, api.getUuid());
+        } catch (GovernanceException e) {
+            log.error(" Error while getting AsyncAPI definition. " + e);
```
Comment (rmsamitha) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:579]: Mention the api name , version, tenent, etc

Comment (rmsamitha) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:579]: Check and fix at other places as well

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java (hunk 10 above)
Comment (HiranyaKavishani) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:414]: just skip revisions

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java [hunk 17]
```diff
@@ -301,6 +339,289 @@ public void registryDataPopulation() throws APIMigrationException {
...
+                        // validate API Definitions
+                        validateAPIDefinitions(api, registry);
+                    } catch (Exception e) {
+                        throw new APIMigrationException("Error occurred while retrieving API from the registry: ", e);
+                    }
+                }
+                log.info("Successfully validated the api definitions of tenant " + tenant.getDomain() + "..........");
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug(
```
Comment (rmsamitha) [inline migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:429]: Isn't this an error than a debug log? Or this this expected sometimes?

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java (hunk 2 above)
Comment (HiranyaKavishani) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:177]: formattedTenantConf  == null only due to L164. So do we need mention it in here?

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java (hunk 17 above)
Comment (HiranyaKavishani) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:427]: No this is not an error. artifactManager with API can be empty. Added this else block to identify only the reason

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java (hunk 12 above)
Comment (HiranyaKavishani) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:480]: L-474 added the api name , version. we don't need it here

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java (hunk 13 above)
Comment (HiranyaKavishani) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:513]: L-507 added the info. we don't need to include them at here

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java (hunk 14 above)
Comment (HiranyaKavishani) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java:559]: L553 added the api info. so we dont need them at here

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java (hunk 1 above)
Comment (HiranyaKavishani) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java]: done

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java (hunk 3 above)
Comment (HiranyaKavishani) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java]: Done

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java (hunk 4 above)
Comment (HiranyaKavishani) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java]: Done

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrationClientBase.java (hunk 6 above)
Comment (HiranyaKavishani) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrationClientBase.java]: Done

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java (hunk 11 above)
Comment (HiranyaKavishani) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java]: done

File: migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java [hunk 18]
```diff
@@ -301,6 +339,289 @@ public void registryDataPopulation() throws APIMigrationException {
...
+        }
+
+        if (api.getGraphQLSchema() == null) {
+            GraphQLSchemaDefinition definition = new GraphQLSchemaDefinition();
+            try {
+                String graphqlSchema = definition.getGraphqlSchemaDefinition(api.getId(), registry);
+                graphQLValidationResponseDTO = PublisherCommonUtils
+                        .validateGraphQLSchema("schema.graphql", graphqlSchema);
+            } catch (APIManagementException e) {
+                log.error(" Error while validating graphql api definition. " + e);
```
[the remark this answers is not in the document]
Comment (HiranyaKavishani) [reply migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/client/MigrateFrom400.java]: done

**Pull request merged**